lib/beacon, lib/discover: Send IPv4 limited broadcast when address listing fails (fixes #1628)

bt90 opened the pull request:
### Purpose

Fixes [#1628](https://github.com/syncthing/syncthing-android/issues/1628)

The basic idea is to continue with local discovery even if we can't list interfaces. Since Android doesn't allow us to collect our own device IPs, we let the receiving clients do the job by advertising unspecified addresses, as we do for global discovery.

Note that this only fixes IPv4 broadcast discovery. IPv6 multicast discovery relies on network interface information and is still broken.

### Testing

Tested on my Pixel 4a 5G. My other client running an unpatched syncthing v1.24.0 successfully resolved the unspecified addresses.

## Authorship

Your name and email will be added automatically to the AUTHORS file
based on the commit metadata.

bt90 commented on lib/beacon/broadcast.go:
```diff
@@ -47,7 +47,6 @@ func writeBroadcasts(ctx context.Context, inbox <-chan []byte, port int) error {
 		intfs, err := net.Interfaces()
 		if err != nil {
 			l.Debugln(err)
-			return err
```
`net.Interfaces()` will throw an error on Android. We can still carry on and use the fallback broadcast address `255.255.255.255` later on.

acolomb replied:
This comment should maybe live directly in the code.

bt90 commented on lib/beacon/broadcast.go:
```diff
@@ -59,7 +58,7 @@ func writeBroadcasts(ctx context.Context, inbox <-chan []byte, port int) error {
 			if err != nil {
 				l.Debugln(err)
-				return err
+				continue
```
Interface discovery might work while retrieving the addresses doesn't. So log the error and carry on.

acolomb replied:
Code comment instead of only GitHub?

bt90 commented on lib/beacon/multicast.go:
```diff
@@ -135,6 +134,17 @@ func readMulticasts(ctx context.Context, outbox chan<- recv, addr string) error
 
+	// Handle failed interface lookup on Android
+	if len(intfs) == 0 {
+		err := pconn.JoinGroup(nil, &net.UDPAddr{IP: gaddr.IP})
```
> JoinGroup uses the system assigned multicast interface when ifi is nil

https://pkg.go.dev/golang.org/x/net/ipv6#PacketConn.JoinGroup

bt90 commented on lib/discover/local.go:
```diff
@@ -292,9 +289,9 @@ func (c *localClient) registerDevice(src net.Addr, device Announce) bool {
-func filterUnspecifiedLocal(addrs []string) []string {
+// filterUndialableLocal returns the list of addresses after removing any
+// localhost, multicast, broadcast or port-zero addresses.
+func filterUndialableLocal(addrs []string) []string {
```
Renamed as we want to keep unspecified addresses, but still want to filter the remaining garbage.

bt90 commented:
Funnily enough, a potential fix for Golang itself seems to be emerging: https://github.com/golang/go/pull/61089

The question is whether it would be accepted and in which go version it would be delivered. The earliest date for a version with the fix would be February 2024 with go v1.22.

Should I continue with my tests, or is this too much of a hack and we'd better wait for a fix at the language level?

imsodin commented:
> [...] we let the receiving clients do the job by advertising unspecified addresses, as we do for global discovery.

I didn't find where wo do that, i.e. where the receiver of broad-/multicast messages resolves the sender address when the advertised one is unspecified?

bt90 commented:
https://github.com/syncthing/syncthing/blob/f25a169c4c0943b34777a306e93d8fbefc9d8109/lib/discover/local.go#L232

bt90 commented:
The logging is a bit borked, but the IPv4 broadcast is properly transmitted and interpreted by my other client:

```
[NTMID] 2023/09/10 18:30:46.215548 local.go:206: DEBUG: discover: Received local announcement from 192.168.178.22:45269 for CG4YCIW
[NTMID] 2023/09/10 18:30:46.215548 local.go:235: DEBUG: discover: Registering addresses for CG4YCIW
[NTMID] 2023/09/10 18:30:46.215548 local.go:269: DEBUG: discover: Reconstructed URL is &url.URL{Scheme:"tcp", Opaque:"", User:(*url.Userinfo)(nil), Host:"192.168.178.22:22000", Path:"", RawPath:"", OmitHost:false, ForceQuery:false, RawQuery:"", Fragment:"", RawFragment:""}
[NTMID] 2023/09/10 18:30:46.215548 local.go:271: DEBUG: discover: Replaced address 0.0.0.0 in tcp://0.0.0.0:22000 to get tcp://192.168.178.22:22000
[NTMID] 2023/09/10 18:30:46.215548 local.go:274: DEBUG: discover: Accepted address relay://88.212.32.226:22067/?id=5YBUXQ6 verbatim
[NTMID] 2023/09/10 18:30:46.215548 local.go:269: DEBUG: discover: Reconstructed URL is &url.URL{Scheme:"quic", Opaque:"", User:(*url.Userinfo)(nil), Host:"192.168.178.22:22000", Path:"", RawPath:"", OmitHost:false, ForceQuery:false, RawQuery:"", Fragment:"", RawFragment:""}
[NTMID] 2023/09/10 18:30:46.215548 local.go:271: DEBUG: discover: Replaced address 0.0.0.0 in quic://0.0.0.0:22000 to get quic://192.168.178.22:22000
[NTMID] 2023/09/10 18:30:46.215548 local.go:274: DEBUG: discover: Accepted address quic://80.xxx.xxx.xxx:61166 verbatim
```

bt90 commented:
Local discovery works over IPv4. Multicast doesn't seem to work unless an interface is explicitly selected. There are no errors logged other than the usual suspects, but I don't see any announcements from my local client.

I also noticed something else while testing this PR. Since interface listing doesn't work, our LAN detection on Android treats everything as WAN. So the address selection is a bit erratic.

bt90 commented:
I reverted the multicast changes as I couldn't get it to work. The local discovery via IPv4 broadcast is fixed though.

calmh commented on lib/beacon/broadcast.go:
```diff
@@ -46,8 +46,9 @@ func writeBroadcasts(ctx context.Context, inbox <-chan []byte, port int) error {
 		if err != nil {
-			l.Debugln(err)
-			return err
+			l.Debugln("Failed to list interfaces", err)
```
```suggestion
			l.Debugln("Failed to list interfaces:", err)
```

calmh commented on lib/beacon/broadcast.go:
```diff
@@ -58,8 +59,9 @@ func writeBroadcasts(ctx context.Context, inbox <-chan []byte, port int) error {
 			if err != nil {
-				l.Debugln(err)
-				return err
+				l.Debugln("Failed to list interface addresses", err)
```
```suggestion
				l.Debugln("Failed to list interface addresses:", err)
```

calmh commented on lib/discover/local.go:
```diff
@@ -310,8 +307,7 @@ func filterUnspecifiedLocal(addrs []string) []string {
 		case tcpAddr.Port == 0:
-		case tcpAddr.IP.IsUnspecified():
-		case !tcpAddr.IP.IsGlobalUnicast() && !tcpAddr.IP.IsLinkLocalUnicast():
+		case !tcpAddr.IP.IsGlobalUnicast() && !tcpAddr.IP.IsLinkLocalUnicast() && !tcpAddr.IP.IsUnspecified():
 		default:
```
```suggestion
		case tcpAddr.IP.IsGlobalUnicast(), tcpAddr.IP.IsLinkLocalUnicast(), tcpAddr.IP.IsUnspecified():
```

let's use a positive expression instead

calmh commented:
Wow, so the "Fixes" in the body there becomes very misleading with GitHub rendering... For the record, we do those things in the commit subject, and none of the tooling supports linking in issues from other repos... 🤷

tomasz1986 commented:
> Wow, so the "Fixes" in the body there becomes very misleading with GitHub rendering... For the record, we do those things in the commit subject, and none of the tooling supports linking in issues from other repos... 🤷

I think I've seen this in other issues too. GitHub really should display the link as `syncthing-android#1628` or similar to avoid the confusion.
